### DISCUSSION TITLE
How can I output just the code from a quarto document as a `.R` File

### DISCUSSION THREAD

**Discussion by ChandlerLutz** [quarto-dev/quarto-cli#7683] [Q&A] [answered]
### Description

I have a standard quarto `.qmd` file. I would like to output the code to a `.R` file. 

- Is it possible to do this? 

- If I wanted to output the code from a python file to a `.py` file, would the process be any different?

**Comment by mcanouil** [comment, marked as answer]
See `knitr::purl` (https://bookdown.org/yihui/rmarkdown-cookbook/purl.html).

**Reply by cderv** [reply]
@ChandlerLutz To be complete regarding your question: 

`purl()` support for .qmd file requires [knitr 1.42](https://github.com/yihui/knitr/blob/master/NEWS.md#changes-in-knitr-version-142) for correct support

And this will only allow to create .R file from a .qmd cell content. 

> If I wanted to output the code from a python file to a .py file, would the process be any different?

Yes it would. You would probably need to leverage [Jupyter CLI](https://jupytext.readthedocs.io/en/latest/using-cli.html#jupytext-cli) to convert to a python py script. 

So overall, Quarto does not have feature to convert to a script. Though in 1.4 pre-release, there is support for creating output from `.py 
https://quarto.org/docs/prerelease/1.4/script.html

.R support coming soon

**Reply by allefeld** [reply]
For Jupyter-based `qmd` files, there's also (now?) the approach to use
```bash
quarto convert document.qmd    # → document.ipynb
```
and then 
```bash
jupyter nbconvert --to python document.ipynb    # → document.py
```
if the language is Python, or
```bash
jupyter nbconvert --to script document.ipynb    # → document.txt
```
otherwise. However, the output from `--to script` is pretty barebones, it doesn't include metadata or Markdown cells.

***

Many more languages are supported by [`jupytext`](https://jupytext.readthedocs.io/en/latest/using-cli.html); it can read `ipynb` files but also directly `qmd` files. For example
```bash
jupytext --to jl document.qmd    # → document.jl
````
for a Julia-based document.

***

Finally, it's pretty easy to process an `ipynb` file (generated by `quarto convert`) from any language which supports reading JSON, with the advantage that detailed customization is possible. I've done that [for Matlab](https://github.com/allefeld/cvcrossmanova/blob/main/quarto-docs/quartoToScript.m).
